### PR TITLE
gifs -> mp4s

### LINKED ISSUE #612: replace large .gifs with smaller .mp4s
- all `.gif` references now use `.mp4` and the `<video>` tag
- videos are lazy loaded (yalljs)  when users scroll near them, and have placeholder images of de-scaled first frame images
- mp4s are in the "snippets" folder - `./source/img/snippets/*.mp4`, along with corresponding `.png` placeholder image
- mp4s were from existing .gifs, except for the following, which came from source (so the following are high quality): 
    - debugging.mp4
    - installing-cli.mp4
    - installing-global.mp4
    - running-tests.mp4
    - writing-tests.mp4

we should probably work on getting higher quality videos for the others

- add `add-snippet` npm script, to take an existing gif or mp4 and put it in the source properly, it makes a placeholder image and runs it through ffmpeg

### PR DESCRIPTION
**Pull request by Bkucera**
- all `.gif` references now use `.mp4` and the `<video>` tag
- videos are lazy loaded (yalljs)  when users scroll near them, and have placeholder images of de-scaled first frame images
- mp4s are in the "snippets" folder - `./source/img/snippets/*.mp4`, along with corresponding `.png` placeholder image
- mp4s were from existing .gifs, except for the following, which came from source (so the following are high quality): 
    - debugging.mp4
    - installing-cli.mp4
    - installing-global.mp4
    - running-tests.mp4
    - writing-tests.mp4

we should probably work on getting higher quality videos for the others

- add `add-snippet` npm script, to take an existing gif or mp4 and put it in the source properly, it makes a placeholder image and runs it through ffmpeg

closes #612

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -33,6 +33,10 @@ gulp.task('move:menu:spy:js', function () {
   return moveJSNodeModule('menuspy/dist/menuspy.js')
 })
 
+gulp.task('move:doc:yall:js', function () {
+  return moveJSNodeModule('yall/dist/yall-2.0.1.min')
```
Comment (jennifer-shehane) [inline gulpfile.js]: Won't this task break on any update of `yall` to another version? Can we glob this `yall/dist/yall*min`? 

Also, no need to move over minified file, as minification is a later step - always prefer non-minified when copying to source for easier viewing in dev.

**Comment by jennifer-shehane** [comment]
The tests are also failing - likely this indicates the root cause.

File: gulpfile.js (hunk 1 above)
Comment (Bkucera) [reply gulpfile.js]: actually, we have to use dist bc it's ES6, and the minification step does not transpile anything

**Comment by jennifer-shehane** [comment]
@Bkucera Tests are still failing:

> Uncaught ReferenceError: yall is not defined

**Comment by jennifer-shehane** [comment]
So, the changes are great. And I was like, ok...but is it really loading faster? 
The answer is yes 😂 😂 😂 😂 😂 😂 

*Before*
fnuljzq0kj

*After*
5zacxbgdf4

File: cy_scripts/snippet-creator/cli.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      debug.enabled = true
+    }
+
+    const [inputFile] = program.args
+    const outBaseFile = `source/img/snippets/${removeExtension(path.basename(inputFile))}`
+    const outFile = path.resolve(`${outBaseFile}.mp4`)
+    const outImg = path.resolve(`${outBaseFile}.png`)
+
+    return Promise.all([fs.pathExists(inputFile), fs.pathExists(outFile)])
+    .then(([inExists, outExists]) => {
```
Comment (brian-mann) [inline cy_scripts/snippet-creator/cli.js]: `.spread`

File: cy_scripts/snippet-creator/convert.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    //ffmpeg -i $filename.mp4 -r 1/1 -vframes 1  $filename.png
+    }),
+  ])
+  .then(() => {
+    spinner.stop()
+    return filesize
+  })
+  .catch((err) => {
+    spinner.stop()
```
Comment (brian-mann) [inline cy_scripts/snippet-creator/convert.js]:
```js
.then(() => {
  return filesize
})
.finally(spinner.stop)
```

File: cy_scripts/snippet-creator/cli.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  },
+}
+
+
+module.exports = {
+  start,
+}
+
+function removeExtension (filename) {
+  return filename.replace(/\.[^/.]+$/, '')
```
Comment (brian-mann) [inline cy_scripts/snippet-creator/cli.js]: `path.extname(...)`

File: cy_scripts/snippet-creator/cli.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  err.known = true
+  throw err
+}
+
+const errors = {
+  inputFileNoExist: (infile) => `Cannot find input file: ${infile}`,
+  badArgs: (num) => `Script takes exactly 1 argument, received ${num}`,
+}
+
+const Log = {
```
Comment (brian-mann) [inline cy_scripts/snippet-creator/cli.js]: `Log` is not a class. It's a utility, should be lowercase.

File: cy_scripts/snippet-creator/cli.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+const chalk = require('chalk')
+const fs = require('fs-extra')
+const debug = require('debug')('snippet')
+const path = require('path')
+const Confirm = require('prompt-confirm')
+
+
+const convert = require('./convert')
+
+global.Promise = require('bluebird')
```
Comment (brian-mann) [inline cy_scripts/snippet-creator/cli.js]: This is almost never a good idea to do. it's generally better to always import it. You could be overriding the way stuff works in code that's not yours that's using native promises. Although bluebird is technically compliant with A+ Promises, it's just not a good idea.

File: cy_scripts/snippet-creator/convert.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  })
+
+
+}
+
+module.exports = {
+  convert,
+}
+
+const ffError = (err, stdout, stderr, reject) => {
```
Comment (brian-mann) [inline cy_scripts/snippet-creator/convert.js]: This is a perfect opportunity to use currying.

```js
const ffError = (reject) => (err, stdout, stderr) => {
  ...
}

// and now you can do just this...
.on('error', => ffError(reject))
```

**Pull request merged**